compiler/hash-typecheck/src/error.rs

```diff
@@ -36,6 +36,43 @@ impl Symbol {
     }
 }
 
+/// This is a generic representation of a class of errors that inform the user that
+/// some construct required an exact amount of arguments but however found a different
+/// amount of arguments. This is used by many variants in the [TypecheckError] enum and
+/// it implements a collection of useful methods for creating report notes/suggestions.
+#[derive(Debug)]
+pub struct ArgumentLengthMismatch {
+    wanted: usize,
+    given: usize,
+}
+
+impl ArgumentLengthMismatch {
+    /// Create a new [ArgumentLengthMismatch]
+    pub fn new(wanted: usize, given: usize) -> Self {
+        ArgumentLengthMismatch { wanted, given }
+    }
+
+    /// Convert the arguments mismatch into a printable reporting note that essentially
+    /// suggests if an argument should be added or removed.
+    pub fn to_note(&self) -> String {
+        let Self { wanted, given } = self;
+
+        if wanted > given {
+            let diff = wanted - given;
+
+            format!("consider adding {} argument(s).", diff)
+        } else {
+            let diff = given - wanted;
+
+            if diff == 1 {
+                "consider removing the last argument.".to_string() // @@Reporting: suggestions!
+            } else {
+                format!("consider removing the last {} arguments.", diff)
+            }
```

Comment:
The wrong argument(s) aren't always going to be the last. Maybe it's better to only say this if the types agree too?

Reply:
I guess we can determine this when we make the error?